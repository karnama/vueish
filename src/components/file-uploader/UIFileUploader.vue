<template>
    <div>
        <div :class="{
                 'active': isDraggedOver && !isLoading,
                 'pointer-events-none': isLoading,
                 error
             }"
             aria-label="File Upload"
             class="rounded-lg drop-zone border-2 border-dashed transition-colors flex flex-wrap items-stretch
                    dark:text-white relative border-current"
             tabindex="0"
             v-bind="$attrs"
             @dragover.prevent="isDraggedOver = true"
             @keydown.enter="openFileBrowser"
             @dragleave="isDraggedOver = false"
             @drop.prevent="addFiles">
            <input ref="fileInput"
                   type="file"
                   hidden
                   multiple
                   :accept="acceptedMimes.join(',')"
                   @change.prevent="addFiles">

            <div class="grow flex flex-col items-center justify-center py-24 px-2 text-center cursor-pointer"
                 @click="openFileBrowser">
                <slot :upload-icon="uploadIcon">
                    <p class="mb-4 text-brand-600" v-html="uploadIcon" />
                    <p class="text-lg">
                        Drag and Drop file
                        <br>
                        or
                    </p>
                    <UIButton theme="brand" class="mt-2">
                        Browse
                    </UIButton>
                </slot>
            </div>
            <div v-show="files.length"
                 class="w-full sm:w-1/2 flex flex-col py-6 px-2
                        justify-between bg-gray-200 dark:bg-gray-600 items-center"
                 @keydown.enter.stop>
                <div class="max-h-64 pl-6 mr-2 pr-4 file-list w-full divide-y">
                    <UIFile v-for="(file, index) in files"
                            :key="index"
                            :file="file"
                            class="py-2"
                            :upload="upload"
                            :upload-on-mounted="uploadAsap"
                            @removed="removeFile" />
                </div>
                <UIButton v-if="!uploadAsap"
                          theme="brand"
                          :loading="isLoading"
                          class="mt-4"
                          tabindex="0"
                          @click="uploadFiles">
                    Upload
                </UIButton>
            </div>

            <UIFadeTransition>
                <div v-if="isLoading" class="absolute top-0 bottom-0 right-0 left-0 bg-gray-400 bg-opacity-50" />
            </UIFadeTransition>
        </div>
        <UIExpandTransition>
            <p v-if="error" class="text-color-error text-sm pt-2">
                {{ error }}
            </p>
        </UIExpandTransition>
    </div>
</template>

<script lang="ts">
import { defineComponent, ref } from 'vue';
import type { PropType } from 'vue';
import UIFile from './UIFile.vue';
import { getIcon } from '@/helpers';
import UIButton from 'components/button/UIButton.vue';
import type { FileError } from 'types';
import UIFadeTransition from 'components/transitions/UIFadeTransition.vue';
import UIExpandTransition from 'components/transitions/UIExpandTransition.vue';
import { error, positiveOptionalNumber } from '@/shared-props';

// todo - UIFile may make the uploader overflow if the side menu is open when using files with long titles
export default defineComponent({
    name: 'UIFileUploader',

    components: { UIExpandTransition, UIFadeTransition, UIFile, UIButton },

    inheritAttrs: false,

    props: {
        /**
         * Accepted mimes in an array format.
         *
         * @default ['*']
         *
         * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/accept
         */
        acceptedMimes: {
            type: Array as PropType<string[]>,
            default: () => ['*']
        },

        /**
         * Maximum number of files to allow to be selected.
         */
        maxFiles: positiveOptionalNumber,

        /**
         * The maximum size allowed for a single file in bytes.
         */
        maxSize: positiveOptionalNumber,

        /**
         * Upload as soon as the file(s)' selected.
         *
         * @default false
         */
        uploadAsap: {
            type: Boolean,
            default: false
        },

        /**
         * The function that uploads the files.
         */
        upload: {
            type: Function as PropType<(files: File[]) => Promise<void>>,
            required: true
        },

        error
    },

    emits: {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        validationError: (_payload: FileError) => true
    },

    setup: (props, ctx) => {
        const uploadIcon = getIcon('upload');

        const fileInput = ref<HTMLInputElement>();
        const isDraggedOver = ref(false);
        const files = ref<File[]>([]);
        const isLoading = ref(false);

        const findFileIndex = (file: File) => {
            return files.value.findIndex(iFile => file.size === iFile.size && file.name === iFile.name);
        };
        const addFiles = (event: DragEvent | InputEvent) => {
            if (isLoading.value) return;

            isDraggedOver.value = false;
            const fileList: FileList = event instanceof DragEvent
                ? event.dataTransfer!.files
                : (event.target as HTMLInputElement).files!;
            // filter out duplicates
            let filteredList: File[] = Array.from(fileList).filter(file => findFileIndex(file) === -1);

            if (fileList.length !== filteredList.length) {
                ctx.emit('validationError',
                    {
                        message: 'Duplicate files not allowed.',
                        files: Array.from(fileList).filter(file => findFileIndex(file) !== -1)
                    } as FileError
                );
            }

            if (typeof props.maxSize === 'number') {
                const tooBigFiles = filteredList.filter(file => file.size > props.maxSize);

                if (filteredList.length && tooBigFiles.length) {
                    ctx.emit('validationError', {
                        message: 'File size exceeds the specified allowed size.',
                        files: tooBigFiles
                    } as FileError);

                    filteredList = filteredList.filter(file => file.size <= props.maxSize);
                }
            }

            if (typeof props.maxFiles === 'number') {
                const remainingSpace = props.maxFiles - files.value.length;
                const allowedItems = filteredList.splice(0, remainingSpace);
                files.value.push(...allowedItems);

                if (filteredList.length) {
                    ctx.emit('validationError', {
                        message: 'Files exceeded specified max count.',
                        files: filteredList
                    } as FileError);
                }
                return;
            }

            files.value.push(...filteredList);
        };
        const removeFile = (file: File) => {
            files.value.splice(findFileIndex(file), 1);
        };
        const openFileBrowser = () => {
            if (isLoading.value) return;

            if (typeof props.maxFiles === 'number' && files.value.length === props.maxFiles) {
                ctx.emit('validationError', {
                    message: 'Maximum number of files already been added.',
                    files: []
                } as FileError);
                return;
            }

            fileInput.value!.click();
        };
        const uploadFiles = async () => {
            isLoading.value = true;
            await props.upload(files.value)
                .then(resp => {
                    files.value = [];
                    return resp;
                })
                .finally(() => {
                    isLoading.value = false;
                });
        };

        return {
            fileInput,
            isDraggedOver,
            files,
            uploadIcon,
            isLoading,
            addFiles,
            removeFile,
            openFileBrowser,
            uploadFiles
        };
    }
});
</script>

<style lang="scss" scoped>
.drop-zone.active {
    background: rgba(0, 0, 0, .1);

    & .file-list {
        pointer-events: none;
}
}

.drop-zone:not(.active) {
    @apply border-gray-300;
}
.error.drop-zone {
    border-color: theme('colors.red.700') !important;
}
.dark .error.drop-zone {
    border-color: theme('colors.red.600') !important;
}

.drop-zone {
    overflow: clip;
}
</style>
