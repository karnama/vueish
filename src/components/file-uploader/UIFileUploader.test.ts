import { mount } from '@vue/test-utils';
import UIFileUploader from './UIFileUploader.vue';

describe('UIFileUploader', () => {
    const upload = () => {};
    const remove = upload;
    it('should display correctly', () => {
        const wrapper = mount(UIFileUploader, {
            props: {
                upload,
                remove
            }
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    // jsdom has no FileList interface implemented
    describe.skip('file validation', () => {
        // const file = new File(['foo'], 'foo.txt', {
        //     type: 'text/plain'
        // });

        // const jsdomUtils = require('jsdom/lib/jsdom/living/generated/utils');
        // const jsdomFileList = require('jsdom/lib/jsdom/living/generated/FileList');
        // function makeFileList(...files: File[]) {
        //     const impl = jsdomFileList.createImpl(window);
        //     const ret = Object.assign([...files], {
        //         item: (ix: number) => ret[ix],
        //         [jsdomUtils.implSymbol]: impl
        //     });
        //     impl[jsdomUtils.wrapperSymbol] = ret;
        //     Object.setPrototypeOf(ret, FileList.prototype);
        //     return ret;
        // }

        // also test the emitted error
        /**
         * @jest-environment ./src/components/file-uploader/file-env
         */
        // can't be spending more time on this
        // eslint-disable-next-line jest/no-disabled-tests
        it.skip('should filter out duplicate files', async () => {
            const wrapper = mount(UIFileUploader, {
                props: {
                    upload,
                    remove
                }
            });

            const input = wrapper.get('input');
            // const data = new DataTransfer();
            // data.items.add(file);
            // data.items.add(file);
            // input.element.files = makeFileList(file, file);
            // Object.defineProperty(input.element, 'files', {
            //     value: [file, file],
            //     writable: true
            // });
            await input.trigger('change');

            expect(wrapper.lastEventValue('validationError')).toStrictEqual([{
                messages: ''
            }]);
        });

        it.todo('should filter out files larger than the given prop');

        it.todo('should not exceed the file count if given as prop');
    });
});
