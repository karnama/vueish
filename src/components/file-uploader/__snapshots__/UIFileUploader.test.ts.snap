// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIFileUploader should display correctly 1`] = `
<div>
  <div
    aria-label="File Upload"
    class="rounded-lg drop-zone border-2 border-dashed transition-colors flex flex-wrap items-stretch
                    dark:text-white relative"
    tabindex="0"
  >
    <input
      accept="*"
      hidden=""
      multiple=""
      type="file"
    />
    <div
      class="flex-grow flex flex-col items-center justify-center py-24 px-2 text-center cursor-pointer"
    >
      <p
        class="mb-4 text-brand-600"
      >
        <svg
          style="width:35px;height:35px"
          viewBox="0 0 24 24"
        >
          <path
            d="M19.35,10.04C18.67,6.59 15.64,4 12,4C9.11,4 6.6,5.64 5.35,8.04C2.34,8.36 0,10.91 0,14A6,6 0 0,0 6,20H19A5,5 0 0,0 24,15C24,12.36 21.95,10.22 19.35,10.04M19,18H6A4,4 0 0,1 2,14C2,11.95 3.53,10.24 5.56,10.03L6.63,9.92L7.13,8.97C8.08,7.14 9.94,6 12,6C14.62,6 16.88,7.86 17.39,10.43L17.69,11.93L19.22,12.04C20.78,12.14 22,13.45 22,15A3,3 0 0,1 19,18M8,13H10.55V16H13.45V13H16L12,9L8,13Z"
            fill="currentColor"
          />
        </svg>
      </p>
      <p
        class="text-lg"
      >
         Drag and Drop file 
      </p>
       or 
      <button
        class="border-brand-400 text-white bg-brand-400 ring-brand-700 hover:bg-brand-500 hover:border-brand-500 border px-3.5 py-2 ui-button rounded relative font-bold text-sm m-0 focus:outline-none ring-0 focus:ring-2 mt-2"
        type="button"
      >
        <transition-stub>
          
          <span
            class="label transition-opacity duration-200"
          >
            
             Browse 
            
          </span>
          
        </transition-stub>
      </button>
    </div>
    <div
      class="w-full sm:w-1/2 flex flex-col py-6 px-2
                        justify-between bg-gray-200 dark:bg-gray-600 items-center"
      style="display: none;"
    >
      <div
        class="overflow-y-scroll max-h-64 pl-6 mr-2 pr-4 file-list space-y-2 w-full"
      >
        
        
      </div>
      <button
        class="border-brand-400 text-white bg-brand-400 ring-brand-700 hover:bg-brand-500 hover:border-brand-500 border px-3.5 py-2 ui-button rounded relative font-bold text-sm m-0 focus:outline-none ring-0 focus:ring-2 mt-4"
        tabindex="0"
        type="button"
      >
        <transition-stub>
          
          <span
            class="label transition-opacity duration-200"
          >
            
             Upload 
            
          </span>
          
        </transition-stub>
      </button>
    </div>
    <transition-stub>
      
      
    </transition-stub>
  </div>
  <transition-stub>
    
    
  </transition-stub>
</div>
`;
